upstream: https://github.com/vim/vim/pull/8180

From 52487608e757f64c137f532b3ded14fb97a62e92 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Thu, 6 May 2021 11:38:53 +0200
Subject: [PATCH] Use POSIX style redirect for ash and dash in shellpipe

Update comment to say "POSIX style redirection" since the `2>&1`
redirection is specified in POSIX and is not a bourne shell specific
thing[1][2].

Add busybox ash and dash to the list of known POSIX compatible shells so
stderr is captured when SHELL is set to one of those.

Add dash.exe to windows but not ash, since busybox ash does not run on
windows.

[1]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_02
[2]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_06
---
 runtime/doc/options.txt | 6 +++---
 src/option.c            | 5 ++++-
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index 84969b23f65..efab64802e5 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -6606,9 +6606,9 @@ A jump table for the options with a short description can be found at |Q_op|.
 	in a file and echoed to the screen.  If the 'shell' option is "csh" or
 	"tcsh" after initializations, the default becomes "|& tee".  If the
 	'shell' option is "sh", "ksh", "mksh", "pdksh", "zsh", "zsh-beta",
-	"bash" or "fish" the default becomes "2>&1| tee".  This means that
-	stderr is also included.  Before using the 'shell' option a path is
-	removed, thus "/bin/sh" uses "sh".
+	"bash", "fish", "ash" or "dash" the default becomes "2>&1| tee".  This
+	means that stderr is also included.  Before using the 'shell' option a
+	path is removed, thus "/bin/sh" uses "sh".
 	The initialization of this option is done after reading the ".vimrc"
 	and the other initializations, so that when the 'shell' option is set
 	there, the 'shellpipe' option changes automatically, unless it was
diff --git a/src/option.c b/src/option.c
index 3c248c19319..7c783757b2d 100644
--- a/src/option.c
+++ b/src/option.c
@@ -910,7 +910,7 @@ set_init_3(void)
 	    }
 	}
 	else
-	    // Always use bourne shell style redirection if we reach this
+	    // Always use POSIX shell style redirection if we reach this
 	    if (       fnamecmp(p, "sh") == 0
 		    || fnamecmp(p, "ksh") == 0
 		    || fnamecmp(p, "mksh") == 0
@@ -919,6 +919,8 @@ set_init_3(void)
 		    || fnamecmp(p, "zsh-beta") == 0
 		    || fnamecmp(p, "bash") == 0
 		    || fnamecmp(p, "fish") == 0
+		    || fnamecmp(p, "ash") == 0
+		    || fnamecmp(p, "dash") == 0
 # ifdef MSWIN
 		    || fnamecmp(p, "cmd") == 0
 		    || fnamecmp(p, "sh.exe") == 0
@@ -929,6 +931,7 @@ set_init_3(void)
 		    || fnamecmp(p, "zsh-beta.exe") == 0
 		    || fnamecmp(p, "bash.exe") == 0
 		    || fnamecmp(p, "cmd.exe") == 0
+		    || fnamecmp(p, "dash.exe") == 0
 # endif
 		    )
 	    {
